File: jzoj/4091.pas
const	maxn=100000;
type	node=record
		next,y,st:longint;
	end;
	arr=array[0..maxn] of boolean;
var	i,j,k,m,n,tot,t,x,y:longint;
	a,e:array[0..maxn] of longint;
	h:array[1..12*maxn] of node;
	s:array[1..6*maxn] of longint;
	p1,p2:arr;
	flag:boolean;
procedure add(x,y,st:longint);
begin
	x:=x mod n;
	y:=y mod n;
	inc(tot);
	h[tot].y:=y;
	h[tot].next:=e[x];
	h[tot].st:=st;
	e[x]:=tot;
end;
procedure bfs(x,st:longint;var p:arr);
var	i,l,r:longint;
begin
	fillchar(p,sizeof(p),0);
	p[x]:=true;
	l:=1;
	r:=1;
	s[l]:=x;
	while l<=r do begin
		i:=e[s[l]];
		while i<>0 do begin
			if (h[i].st=st)and(not p[h[i].y]) then begin
				p[h[i].y]:=true;
				inc(r);
				s[r]:=h[i].y;
			end;
			i:=h[i].next;
		end;
		inc(l);
	end;
end;
begin
	assign(input,'4091.in');reset(input);
	//assign(output,'.out');rewrite(output);

	readln(t);
	while t>0 do begin
		fillchar(a,sizeof(a),255);
		fillchar(e,sizeof(e),0);
		tot:=0;

		readln(n,m);
		for i:=1 to m do begin
			readln(x,y);
			a[x]:=y;
		end;
		for i:=0 to n-2 do begin
			for j:=i+1 to i+6 do begin
				if a[j]<>-1 then begin
					add(i,a[j],1);
					add(a[j],i,2);
				end else begin
					add(i,j,1);
					add(j,i,2);
				end;
			end;
		end;
		bfs(0,1,p1);
		bfs(n-1,2,p2);
		flag:=false;
		for i:=0 to n-1 do
			if p1[i] and (not p2[i]) then begin
				flag:=true;
				break;
			end;
		if flag then writeln('YES') else writeln('NO');

		dec(t);
	end;
end.

File: OI/jzoj/4091.cpp
#include <cstdio>
#include <cstring>
#define N 100000

struct node{
	int next,y,st;
}h[12*N];

int a[N],e[N],s[6*N],n,m,tot;
bool p1[N],p2[N],flag;

void add(int x,int y,int st)
{
	x=x%n;
	y=y%n;
	tot++;
	h[tot].y=y;
	h[tot].next=e[x];
	h[tot].st=st;
	e[x]=tot;
}

void bfs1(int x,int st)
{
	int i;
	memset(p1,0,sizeof p1);
	p1[x]=true;
	int l=1,r=1;
	s[l]=x;
	while (l<=r) {
		i=e[s[l]];
		while (i!=0)
		{
			if ((h[i].st==st) && (!p1[h[i].y])){
				p1[h[i].y]=1;
				r++;
				s[r]=h[i].y;
			}
			i=h[i].next;
		}
		l++;
	}

}
void bfs2(int x,int st)
{
	int i;
	memset(p2,0,sizeof p2);
	p2[x]=true;
	int l=1,r=1;
	s[l]=x;
	while (l<=r) {
		i=e[s[l]];
		while (i!=0)
		{
			if ((h[i].st==st) && (!p2[h[i].y])){
				p2[h[i].y]=1;
				r++;
				s[r]=h[i].y;
			}
			i=h[i].next;
		}
		l++;
	}

}
int main()
{
	freopen("4091.in","r",stdin);
	
	int t;
	scanf("%d\n",&t);
	for (;t>0;t--)
	{
		memset(a,255,sizeof a);
		memset(e,0,sizeof e);
		tot=0;

		scanf("%d%d",&n,&m);
		for (int i=1;i<=m;i++)
		{
			int x,y;
			scanf("%d%d",&x,&y);
			a[x]=y;
		}
		for (int i=0;i<=n-2;i++)
		{
			for (int j=i+1;j<=i+6;j++)
			{
				if (a[j]!=-1){
					add(i,a[j],1);
					add(a[j],i,2);
				} else {
					add(i,j,1);
					add(j,i,2);
				}
			}
		}
		bfs1(0,1);
		bfs2(n-1,2);
		flag=false;
		for (int i=0;i<=n-1;i++)
			if (p1[i] && (!p2[i])) {
				flag=true;
				break;
			}
		if (flag) printf("YES\n");else printf("NO\n");

	}
}
